official/rnn_ptb/run_training.py

```diff
@@ -0,0 +1,244 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Train a RNN model on the PTB dataset."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import os
+import shutil
+
+import numpy as np
+import tensorflow as tf
+
+import model
+import model_params
+import util
+
+# Create argument parser. Add the help option only when running this file.
+parser = argparse.ArgumentParser(add_help=(__name__ == '__main__'))
+parser.add_argument(
+    '--data_dir',
+    type=str,
+    default='',
+    help='The directory where train, validation, and test data are saved.')
+parser.add_argument(
+    '--model',
+    type=str,
+    default='medium',
+    help='The model configuration. Options: small, medium(default), large')
+parser.add_argument(
+    '--model_dir',
+    type=str,
+    default='/tmp/ptb_model',
+    help='The directory where the model will be saved.')
+parser.add_argument(
+    '--reset_training',
+    action='store_true',
+    help='Add this option to clear the model directory prior to training.')
+FLAGS, unparsed = parser.parse_known_args()
+
+# Set the train, validation, and test file paths as constants.
+TRAIN_FILE = os.path.join(FLAGS.data_dir, 'ptb.train.txt')
+VALID_FILE = os.path.join(FLAGS.data_dir, 'ptb.valid.txt')
+TEST_FILE = os.path.join(FLAGS.data_dir, 'ptb.test.txt')
+
+# Hard coded number of words in the training file (used to calculate the maximum
+# number steps to train the model).
+TRAIN_FILE_WORDS = 929589
+
+
+def input_fn(input_file, vocab_dict, num_epochs=1):
+  """Generates batches of data from the input file."""
+  def _input_fn(params):
+    batch_size, unrolled_count = params.batch_size, params.unrolled_count
+
+    def batch_data(data, steps_per_epoch):
```

Comment:
Nesting functions three levels deep seems to introduce unnecessary complexity; could just be my misunderstanding, but any reason not to extract these and have them take the args they need? Is there an advantage to having them defined in the _input_fn?

Reply:
I originally had them under `_input_fn` for convenience. I agree that 3 nested functions is a bit too much, so I've moved them out.